runtime/vm/LayoutFFITypeHelpers.cpp

```diff
@@ -100,31 +101,93 @@ LayoutFFITypeHelpers::getArrayFFIType(char **layout, UDATA nElements) {
+	PORT_ACCESS_FROM_JAVAVM(_vm);
+
+	if (NULL == _vm->layoutStrFFITypeTable) {
+		_vm->layoutStrFFITypeTable = createLayoutStrFFITypeTable(_vm);
+		if (NULL == _vm->layoutStrFFITypeTable) {
+			goto freeAllMemoryThenExit;
+		}
+	}
+
+	layoutStrFFITypeEntry = (J9LayoutStrFFITypeEntry *)j9mem_allocate_memory(sizeof(J9LayoutStrFFITypeEntry), J9MEM_CATEGORY_VM_FFI);
+	if (NULL == layoutStrFFITypeEntry) {
+		goto freeAllMemoryThenExit;
+	}
+
+	layoutStrFFITypeEntry->layoutStrLength = getLengthOfStructLayout(currentLayout);
+	layoutStrFFITypeEntry->layoutStr = (U_8 *)j9mem_allocate_memory(layoutStrFFITypeEntry->layoutStrLength + 1, J9MEM_CATEGORY_VM_FFI);
+	if (NULL == layoutStrFFITypeEntry->layoutStr) {
+		goto freeAllMemoryThenExit;
+	}
+
+	memcpy(layoutStrFFITypeEntry->layoutStr, currentLayout, layoutStrFFITypeEntry->layoutStrLength);
+	layoutStrFFITypeEntry->layoutStr[layoutStrFFITypeEntry->layoutStrLength] = '\0';
+
+	/* Search the hashtable for the ffi_type if the same struct layout string exists. */
+	resultEntry = findLayoutStrFFIType(_vm->layoutStrFFITypeTable, layoutStrFFITypeEntry);
+	if (NULL != resultEntry) {
+		j9mem_free_memory(layoutStrFFITypeEntry->layoutStr);
+		j9mem_free_memory(layoutStrFFITypeEntry);
+		/* Return the duplicate ffi_type for the same struct if found in the hashtable. */
+		typeFFI = (ffi_type *)(resultEntry->structFFIType);
+		/* Reach the end of the nested struct layout string so as to move to
+		 * the next element in getStructFFITypeElements() if exists.
+		 */
+		(*layout) += resultEntry->layoutStrLength - 1;
+		goto done;
+	}
+
+	/* Skip over the '#' case. */
+	(*layout) = currentLayout + 1;
+
 	/* The struct layout is incremented for the '[' case in getStructFFITypeElements(), in which case
 	 * the remaining layout string is then traversed when getStructFFITypeElements() is called recursively.
 	 */
-	(*layout) += 1; // skip over the '#' case
-	ffi_type **structElements = getStructFFITypeElements(layout);
-	PORT_ACCESS_FROM_JAVAVM(_vm);
+	structElements = getStructFFITypeElements(layout);
+	if (NULL == structElements) {
+		goto freeAllMemoryThenExit;
+	}
+
 	typeFFI = (ffi_type *)j9mem_allocate_memory(sizeof(ffi_type), J9MEM_CATEGORY_VM_FFI);
 	if (NULL == typeFFI) {
-		setNativeOutOfMemoryError(_currentThread, 0, 0);
-		goto done;
+		goto freeAllMemoryThenExit;
 	}
 	typeFFI->size = 0;
 	typeFFI->alignment = 0;
 	typeFFI->type = FFI_TYPE_STRUCT;
 	typeFFI->elements = structElements;
 
+	/* Add the created ffi_type for the struct to the hashtable
+	 * if the corresponding layout string doesn't exist.
+	 */
+	layoutStrFFITypeEntry->structFFIType = (void *)typeFFI;
+	resultEntry = addLayoutStrFFIType(_vm->layoutStrFFITypeTable, layoutStrFFITypeEntry);
+	if (NULL == resultEntry) {
+		freeStructFFIType(typeFFI);
+		goto freeAllMemoryThenExit;
+	}
+
 done:
 	return typeFFI;
+
+freeAllMemoryThenExit:
+	j9mem_free_memory(layoutStrFFITypeEntry->layoutStr);
+	j9mem_free_memory(layoutStrFFITypeEntry);
```

Comment:
This must check for `NULL` to be sure it's safe to access `layoutStr`:
```c
	if (NULL != layoutStrFFITypeEntry) {
		j9mem_free_memory(layoutStrFFITypeEntry->layoutStr);
		j9mem_free_memory(layoutStrFFITypeEntry);
	}
```

Reply:
Fixed.